File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.java
class Solution {
    int left, maxLen;
    
    public String longestPalindrome(String s) {
		int len = s.length();
		// 한자리 단어라면 팰린드롬 무조건 성립 및 가장 긴 케이스
		if (len < 2)
			return s;

		for (int i = 0; i < len - 1; i++) {
			// 짝수 팰린드롬 확인
			extendPalindrome(s, i, i + 1);
			// 홀수 팰린드롬 확인
			extendPalindrome(s, i, i + 2);
		}

		return s.substring(left, left + maxLen);

	}

	public void extendPalindrome(String s, int start, int end) {
		// 투 포인터에 대해서 유효한 범위 인지,
		// 양쪽 끝 문자가 일치하는 경우에 팰린드롬 범위를 앞뒤로 한칸씩 확장
		while (start >= 0 && end < s.length() && s.charAt(start) == s.charAt(end)) {
			start--;
			end++;
		}

		if (maxLen < end - start - 1) {
			left = start + 1;
			maxLen = end - start - 1;
		}
	}
}

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.kt
class Solution {
    var left: Int = 0
    var maxLen: Int = 0
    fun longestPalindrome(s: String): String {
        // 문자 길이 저장
        val len = s.length

        // 길이가 1인 경우 예외 처리
        if (len < 2) return s

        // 우측으로 한칸씩 이동하면 투 포인터 조사
        for (i in 0 until len - 1) {
            extendPalindrome(s, i, i + 1)
            extendPalindrome(s, i, i + 2)

        }

        return s.substring(left, left + maxLen)
    }

    fun extendPalindrome(s: String, start: Int, end: Int) {
        // 파라미터 수정 불가 별도 var 선언
        var l = start
        var r = end

        // 투 포인터가 유효한 범위 내 있고 양쪽 끝 문자가 팰린드롬인 경우
        while (l >= 0 && r < s.length && s[l] == s[r]) {
            l--
            r++
        }

        if (maxLen < r - l - 1) {
            left = l + 1
            maxLen = r - l - 1
        }
    }
}